pennylane_lightning/core/src/simulators/lightning_tensor/cutn/MPSCutn.hpp

```diff
@@ -0,0 +1,293 @@
+  private:
+    using BaseType = CutnBase<Precision, MPSCutn>;
+
+    std::size_t maxBondDim_;
+
+    std::vector<std::vector<std::size_t>> sitesModes_;
+    std::vector<std::vector<std::size_t>> sitesExtents_;
+
+    std::vector<std::vector<int64_t>> sitesExtents_int64_;
+    std::vector<int64_t *> sitesExtentsPtr_int64_;
+    std::vector<void *> tensorsDataPtr_;
+    std::vector<CudaTensor<Precision>> tensors_;
+
+  public:
+    using CFP_t = decltype(cuUtil::getCudaType(Precision{}));
+    using ComplexT = std::complex<Precision>;
+
+  public:
+    MPSCutn() = delete;
+
+    explicit MPSCutn(const std::size_t numQubits, const std::size_t maxBondDim)
+        : BaseType(numQubits), maxBondDim_(maxBondDim) {
+        initHelper_();
+    }
+
+    explicit MPSCutn(const std::size_t numQubits, const std::size_t maxBondDim,
+                     DevTag<int> &dev_tag)
+        : BaseType(numQubits, dev_tag), maxBondDim_(maxBondDim) {
+        initHelper_();
+    }
+
+    ~MPSCutn() = default;
+
+    /**
+     * @brief Get the max bond dimension.
+     *
+     * @return std::size_t
+     */
+    [[nodiscard]] auto getMaxBondDim() const -> std::size_t {
+        return maxBondDim_;
+    };
+
+    /**
+     * @brief Get a vector of pointers to extents of each site
+     *
+     * @return sitesExtentsPtr_int64_ std::vector<int64_t *> Note int64_t is
+     * required by cutensornet backend.
+     */
+    [[nodiscard]] auto getSitesExtentsPtr() -> std::vector<int64_t *> & {
+        return sitesExtentsPtr_int64_;
+    }
+
+    /**
+     * @brief Get a vector of pointers to tensor data of each site
+     *
+     * @return tensorsDataPtr_ std::vector<void *> Note void is required by
+     * cutensornet backend.
+     */
+    [[nodiscard]] auto getTensorsDataPtr() -> std::vector<void *> & {
+        return tensorsDataPtr_;
+    }
+
+    /**
+     * @brief Set a zero state
+     */
+    void reset() {
+        const std::vector<std::size_t> zeroState(BaseType::getNumQubits(), 0);
+        setBasisState(zeroState);
+    }
+
+    /**
+     * @brief Set basis state
+     * NOTE: This API assumes the bond vector is [1,0,0,......] and current
+     * implementation only works for qubits.
+     * @param basisState Vector representation of a basis state.
+     */
+    void setBasisState(const std::vector<std::size_t> &basisState) {
+        PL_ABORT_IF(BaseType::getNumQubits() != basisState.size(),
+                    "The size of a basis state should be equal to the number "
+                    "of qubits.");
+
+        bool allZeroOrOne = std::all_of(
+            basisState.begin(), basisState.end(),
+            [](std::size_t bitVal) { return bitVal == 0 || bitVal == 1; });
+
+        PL_ABORT_IF(allZeroOrOne == false,
+                    "Please ensure all elements of a basis state should be "
+                    "either 0 or 1.");
+
+        PL_ABORT_IF(this->MPSInitialized_ == true,
+                    "setBasisState() can be called only once.");
```

Comment:
It would be nice to get rid of this limitation. If we make `initHelper_` the following allows calling `setBasisState` in a loop.
```suggestion
        if (this->MPSInitialized_){
            PL_CUTENSORNET_IS_SUCCESS(cutensornetStateFinalizeMPS(                        
            /*const cutensornetHandle_t*/ BaseType::getCutnHandle(),
            /*cutensornetState_t*/ BaseType::getQuantumState(),
            /*cutensornetBoundaryCondition_t*/ CUTENSORNET_BOUNDARY_CONDITION_OPEN,
            /*const int64_t *const*/ getSitesExtentsPtr().data(),
            /*const int64_t *const*/ nullptr));
            this->MPSInitialized_ = false;
            this-> MPSFinalized_ = true;
        }
        if (this-> MPSFinalized_){
            BaseType::initHelper_();
            this->MPSFinalized_ = false;
        }
        PL_ABORT_IF(this->MPSInitialized_ == true,
                    "setBasisState() can be called only once.");
```

Reply:
Thanks @vincentmr ! This implementation can get C++ unit test for multiple `reset()` and `setBasisState()` passed.  But my concern is that this could bring kind of wild/dangling pointer issue if we are trying to call `BaseType::initHelper_()` here since `quantumState_` previously created is not properly released. I would like to add a //TODO to come back to it later and move the code blocks in `initHelper_` to the `ctor`. What are your thoughts @vincentmr @mlxd @maliasadi ?

Reply:
Personally, I wouldn't  call `cutensornetDestroyState()` manually since it's not a safe practice. I would say this is a constraint from the `cutensornet` lib and I'd suggest that we can communicate with `cutensornet` team and ask them to lift this constraint for the root.

Reply:
Isn't `cutensornetDestroyState` the safest way to release `quantumState_`? Or what did you mean by 
> this could bring kind of wild/dangling pointer issue if we are trying to call BaseType::initHelper_() here since quantumState_ previously created is not properly released.

Reply:
- Manually call `cutensornetDestroyState()` in the class methods wouldn't be safe because of the lifetime issue. We should ensure that  `quantumState_` has the same lifetime with the graph and other data it associated. To ensure this, we put the `cutensornetDestroyState` into the destructor of the `CuTNBase` class.

- If we call `initHelper_()` in the class methods other than ctor, it will create another `quantumState_` and assign the pointer to the newly created quantumState to  the `quantumState_`. In this case, if we destroy the previous state in advance, there will be a lifetime issue. If we don't, the pointer to the previously created `quantumState_` is lost.

- I also believe that the `reset()` and `setBasisState()` method is to update the existing quantum state instead of destroying the old quantum state and creating a new quantum state. The quantum state should be only created and destroyed by creating and destroying a `MPSCuTN` object.

We can have a call to discuss if needed.

Reply:
Thanks. To be clear, creating a new quantum state and destroying the old one doesn't follow the `reset()` used in our workflows. [`resetStateVector()` ](https://github.com/PennyLaneAI/pennylane-lightning/blob/cde30161c8ee9149d7b9fbbc51a50042a2a93bc6/pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp#L168-L172) in our `Kokkos` backends only update the state vector data instead of deallocating the existing one and allocating a new one.